### PR TITLE
PP-11246 Set mandatory fields for events

### PR DESCRIPTION
**Pull request by james-peacock-gds**
- Make timestamp, externalId, resourceType and eventType mandatory for post events endpoint
- update openapi spec

### CODE REVIEW

File: src/main/java/uk/gov/pay/ledger/queue/EventMessageDto.java [hunk 1]
```diff
@@ -21,19 +22,24 @@ public class EventMessageDto {
 
     private Boolean live;
 
+    @NotNull
```
Comment (kbottla) [inline src/main/java/uk/gov/pay/ledger/queue/EventMessageDto.java:25]: Can we add test(s) for these changes?

Comment (james-peacock-gds) [reply src/main/java/uk/gov/pay/ledger/queue/EventMessageDto.java:25]: have added an integration test to demonstrate that a request missing one of the mandatory fields will be rejected with 422. Is this sufficient?

File: src/test/java/uk/gov/pay/ledger/event/resource/EventResourceAgreementIT.java [hunk 2]
```diff
@@ -85,4 +76,45 @@ public void shouldWriteCancelledEvent() {
...
+                        .put("cancelled_date", now.toString()))
+                .put("resource_external_id", agreementId);
+        params.put(event);
+
+        given().port(port)
+                .contentType(JSON)
+                .body(params.toString())
+                .post("/v1/event")
+                .then()
+                .statusCode(422);
```
Comment (kbottla) [inline src/test/java/uk/gov/pay/ledger/event/resource/EventResourceAgreementIT.java:104]: I think testing 422 here is fine but not clear which field is supposed to be mandatory as per the test name. 

Would be good to update the test name or add a few more assertions for the expected fields without relying on the comment.

File: src/test/java/uk/gov/pay/ledger/event/resource/EventResourceAgreementIT.java [hunk 3]
```diff
@@ -85,4 +76,45 @@ public void shouldWriteCancelledEvent() {
         assertThat(agreementEntity.getCancelledByUserEmail(), is("jdoe@example.org"));
         assertThat(agreementEntity.getCancelledDate(), is(ZonedDateTimeMatchers.within(1, ChronoUnit.SECONDS, now)));
     }
+
+    @Test
+    public void shouldNotWriteEventIfAMandatoryFieldIsNotPresent() {
```
Comment (kbottla) [inline src/test/java/uk/gov/pay/ledger/event/resource/EventResourceAgreementIT.java:81]: This doesn't cover fields other than event_type. We could add a few unit tests for other fields. 

A good place to add unit tests is https://github.com/alphagov/pay-ledger/blob/master/src/test/java/uk/gov/pay/ledger/event/resource/EventResourceTest.java#L24

Comment (kbottla) [inline src/test/java/uk/gov/pay/ledger/event/resource/EventResourceAgreementIT.java:81]: Just noticed that this is in `EventResourceAgreementIT`. This test class is specifically testing for agreement projection. We can add a test to `EventResourceIT`.

And looks like there is already a test which could be improved https://github.com/alphagov/pay-ledger/blob/10b3c7dfb45475395bd6745dbd514dc8e47008a6/src/test/java/uk/gov/pay/ledger/event/resource/EventResourceIT.java#L62 as doesn't verify status or error message being returned.

**Pull request closed without merging**